oca/api.py

```diff
@@ -49,15 +32,66 @@ def method(self):
 
             [1, 2, 3]
 
+        If no container is provided as an argument to the decorator, and a
+        value is returned from the decorated method, a ValueError will be
+        raised.
+
+        .. code-block:: python
+
+            @api.foreach
+            def method(self):
+                # Raises ValueError
+                return 1+2
+
+            @api.foreach
+            def method(self):
+                # No error raised
+                self.math = 1+2
+
+    Args:
+        container (type): Type of container to wrap results in. It will be
+            instantiated with an iterable of the decorated method results.
+
+    Raises:
+        ValueError: If no container type is provided, and a value is returned
+            from the decorated method.
+        TypeError: If an incorrect data type is provided as the `container`.
+            Valid arguments are `None`, instances of `type`, or a callable.
+
     Returns:
         list: Of the return values from the decorated method, in the same
             order as the original ``self`` RecordSet.
     """
 
-    def loop(method, self, *args, **kwargs):
-        result = [method(rec, *args, **kwargs) for rec in self]
-        return aggregate(method, result, self)
-
-    wrapper = decorator(loop, method)
-    wrapper._api = 'foreach'
-    return wrapper
+    method = None
+
+    if container is not None and not isinstance(container, type):
+        if callable(container):
+            method = container
+            container = None
+        else:
+            raise TypeError(
+                'An incorrect argument was provided to the `foreach` api.',
+            )
+
+    def _foreach(method):
+
+        def loop(method, self, *args, **kwargs):
+            results = [method(rec, *args, **kwargs) for rec in self]
+            if container is None:
+                if not all(r is None for r in results):
```

Comment:
I was telling you about `assert` because I think it is a good place to use it.

[Assertions are removed when you Pycompile code with `PYTHONOPTIMIZE=2`](https://docs.python.org/2.7/reference/simple_stmts.html?highlight=assert#the-assert-statement). This means that if you use that feature in your production environments, [like I do in docker-odoo-base](https://github.com/Tecnativa/docker-odoo-base/blob/master/build.d/50-compile), you will save several CPU cycles each time a method with this decorator is called, because it is supposed that you removed the `AssertionError`s when you were developing, and it is supposed that even if any method returns anything, it should not break production.